crates/networking/p2p/sync.rs

```diff
@@ -114,48 +128,57 @@ impl SyncManager {
-                    bytecode_receiver,
-                    self.peers.clone(),
-                    store.clone(),
-                ));
-                set.spawn(fetch_blocks_and_receipts(
+                // - Fetch each block's body and its receipt via eth p2p requests
+                // - Fetch the pivot block's state via snap p2p requests
+                // - Execute blocks after the pivote (like in full-sync)
+                let fetch_bodies_handle = tokio::spawn(fetch_block_bodies(
                     all_block_hashes.clone(),
                     self.peers.clone(),
-                    store.clone(),
-                ));
-                let state_roots = all_block_headers
-                    .iter()
-                    .map(|header| header.state_root)
-                    .collect::<Vec<_>>();
-                set.spawn(fetch_snap_state(
-                    bytecode_sender,
-                    state_roots.clone(),
-                    self.peers.clone(),
-                    store.clone(),
                 ));
-                // Store headers
-                let mut latest_block_number = 0;
-                for (header, hash) in all_block_headers
-                    .into_iter()
-                    .zip(all_block_hashes.into_iter())
-                {
-                    latest_block_number = header.number;
-                    store.set_canonical_block(header.number, hash)?;
-                    store.add_block_header(hash, header)?;
+                let mut pivot_idx = if all_block_headers.len() > MIN_FULL_BLOCKS {
+                    all_block_headers.len() - MIN_FULL_BLOCKS
+                } else {
+                    all_block_headers.len() - 1
+                };
+                let mut pivot_root = all_block_headers[pivot_idx].state_root;
+                let mut pivot_number = all_block_headers[pivot_idx].number;
+
+                let mut stale_pivot =
+                    !rebuild_state_trie(pivot_root, self.peers.clone(), store.clone()).await?;
+                // If the pivot became stale, set a further pivot and try again
+                if stale_pivot && pivot_idx != all_block_headers.len() - 1 {
+                    warn!("Stale pivot, switching to newer head");
+                    pivot_idx = all_block_headers.len() - 1;
+                    pivot_root = all_block_headers[pivot_idx].state_root;
+                    pivot_number = all_block_headers[pivot_idx].number;
+                    stale_pivot =
+                        !rebuild_state_trie(pivot_root, self.peers.clone(), store.clone()).await?;
                 }
-                // If all processes failed then they are likely to have a common cause (such as unaccessible storage), so return the first error
-                for result in set.join_all().await {
-                    result?;
+                if stale_pivot {
+                    warn!("Stale pivot, aborting sync");
+                    return Ok(());
```

Comment:
return Error here?

Reply:
I looked at the error handling on geth and they don't return an error from the main sync process.
The sync process happenes on its own, there is no one we need to reply to with the status of the sync, only the owner/user of the node itself (via tracing) so it doesn't make much sense to return an error unless we want to shut down the node entirely

Reply:
Currently, the rpc context has a mutex reference to the SyncManager, if it can't get hold of the Mutex then it assumes there is an active sync process and returns the syncing status.